src/Umbraco.Web/Trees/ContentTreeControllerBase.cs

```diff
@@ -283,22 +283,35 @@ protected sealed override TreeNodeCollection GetTreeNodes(string id, FormDataCol
         /// This is required in case a user has custom start nodes that are children of a list view since in that case we'll need to render the tree node. In normal cases we don't render
         /// children of a list view.
         /// </remarks>
-        protected bool ShouldRenderChildrenOfContainer(IUmbracoEntity e)
+        protected int ShouldRenderChildrenOfContainer(IUmbracoEntity e)

```

Comment:
Unfortunately this is a breaking change that we shouldn't introduce. This is one of those 'fun' things we need to consider when working on a public product that we don't break APIs. Since this in `protected` it means that users of our software could be using this API.

If we are to change this response, another method should be created with a different name and this one can be marked with `[Obsolete("...")]` with a message explaining to use the new one. The existing method will still need to return a `bool` ... which could wrap your new method. Using an INT return here isn't the clearest approach from an API usage perspective since the caller of this method needs to know exactly what any response might mean which you can't really know by just looking at the method or inspecting the result. One solution could be to use an Enum response.  Based on your comments it would have 4 states:

* HasChildren
* NoChildren
* HasChildrenListView
* NoChildrenListView

That said, it's sort of a combination that we're looking for here: HasChildren, NoChildren, HasChildren | ListView, NoChildren | ListView . It seems like we have this information already so I think we could still just leave this method to return a `bool` and use a combination of the info we've already got:

* `e.HasChildren()`  -> The result from this can be assigned directly to `hasChildren`
* `e.IsContainer()` -> The result from this can be assigned directly to `isContainer`
* `ShouldRenderChildrenOfContainer` -> The `bool` result from this is purely used in the `GetTreeNodesInternal` method

I think that should solve the issue but there's a catch here, one reason why the `ShouldRenderChildrenOfContainer` method was previously used to assign the `hasChildren` was due to the comment you see https://github.com/umbraco/Umbraco-CMS/pull/2859/files#diff-b59f63c97f3f40ddb4f2660b6bcdb42dR305 since IIRC the `hasChildren` property in angular also controls if the node will show children or not so i think that needs to be tested.